lib/player/smart_computer.rb

```diff
@@ -0,0 +1,87 @@
+require_relative '../Token'
+require_relative '../board_transformer'
+
+class SmartComputer
+    include Token
+    include BoardTransformer
+    attr_accessor :token, :winning_token_count
+
+    def move(board:, presenter:)
+        @board = board
+        @transformed_boards = BoardTransformer.transform_with_index(board)
+        @winning_token_count = board.size 
+
+        take_winning || 
+        take_blocking(row_token_count: winning_token_count - 1) || 
+        take_center || 
+        take_empty_corner || 
+        take_row_completing_move(row_token_count: winning_token_count - 2) || 
+        take_blocking(row_token_count: winning_token_count - 2) || 
+        take_random_position
+    end
+
+    private
+    attr_accessor :transformed_boards, :board
+
+    def take_winning
+        find_optimal_move(token_to_find: token, target_token_count: winning_token_count - 1)
```

Comment:
This target_token_count is a bit confusing. It may be simpler to just try all available moves and see if any cause a win. For block you can try all available moves with the opponent's token and see if any cause a win. This count is an interesting idea though!

Reply:
This part has been really hard to name , which certainly doesn't help the confusion.
Essentially, if its blocking, for example:
- first it will block moves that is 1 away from a winning token count (for example, needs 4 adjacent tokens to win)
- then when it runs again on line 19, it will try to find a blocking move within a row that is 2 away from a winning token count.

same algorithm for the winning / row_completing moves, but I had wrapped it in named functions in hopes of clarifying the intention.

Reply:
Maybe `tokens_until_win`?